File: oop/rich/src/main/java/de/hhu/rich/warenkorb/domain/model/common/preis/Preis.java
package de.hhu.rich.warenkorb.domain.model.common.preis;

import java.math.BigDecimal;

public record Preis(BigDecimal betrag, String waehrung) {
    public Preis {
        if (betrag.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException("Betrag darf nicht negativ sein");
        }
    }

    public Preis erhoeheUm(Preis betrag) {
        return new Preis(this.betrag.add(betrag.betrag), betrag.waehrung());
    }

    public Preis reduziereUm(Preis betrag) {
        return new Preis(this.betrag.subtract(betrag.betrag), betrag.waehrung());
    }

    public boolean istGroesserAls(Preis maxEinkaufswert) {
        return betrag.compareTo(maxEinkaufswert.betrag) > 0;
    }
}

File: fp/rich/src/main/clojure/de/hhu/rich/warenkorb/domain/model/preis/Preis.clj
(ns de.hhu.rich.warenkorb.domain.model.preis.Preis)

(defprotocol IPreis
  (erhoehe-um [this betrag]
    "Erhöht den Preis um den angegebenen Betrag.")
  (reduziere-um [this betrag]
    "Reduziert den Preis um den angegebenen Betrag.")
  (ist-groesser-als? [this max-preis]
    "Prüft, ob der Preis größer als der angegebene Maximalpreis ist."))

(defrecord Preis [betrag waehrung]
  IPreis
  (erhoehe-um [this betrag]
    (if (<= betrag BigDecimal/ZERO)
      (throw (IllegalArgumentException. "Die Erhöhung muss größer als 0 sein"))
      (->Preis (+ betrag (:betrag this)) (:waehrung this))))

  (reduziere-um [this betrag]
    (if (<= betrag BigDecimal/ZERO)
      (throw (IllegalArgumentException. "Die Reduzierung muss größer als 0 sein"))
      (->Preis (- (:betrag this) betrag) (:waehrung this))))

  (ist-groesser-als? [this max-preis]
    (> (.compareTo (:betrag this) (:betrag max-preis)) 0)))

(defn erstelle-preis [betrag waehrung]
  (if (neg? (.compareTo betrag BigDecimal/ZERO))
    (throw (IllegalArgumentException. "Der Betrag darf nicht negativ sein"))
    (->Preis betrag waehrung)))
